Implement the schema subcommand from issue #444

KSchopmeyer opened the pull request:
Implements the schema subcommand for the server group.

KSchopmeyer commented:
Example of the current table output from running the command against OpenPegasus:

```

|WARNING: Invalid version_qualifier value 2.10, class CIM_DatabaseControlFile.  Modified to [2, 10, 0]
Schema information
Namespace                      schema version    experimental      count  schemas
-----------------------------  ----------------  --------------  -------  ------------------------------------------------------------------------------
root/PG_InterOp                2.40.0            False               276  PG,CIM
root/benchmark                 2.40.0            False               206  CIM
root/SampleProvider            2.41.0            True                466  PG,CIM,CWS,Sample,CQL,CMPISample,CMPI,TestCMPI,QExpr
test/CimsubTestNS2             2.40.0            False               212  CIM,PG
test/CimsubTestNS3             2.40.0            False               212  CIM,PG
test/CimsubTestNS0             2.40.0            False               212  CIM,PG
test/CimsubTestNS1             2.40.0            False               212  CIM,PG
root/PG_Internal               2.11.0            False                10  PG
test/WsmTest                   2.40.0            False               182  CIM
test/TestIndSrcNS1             2.40.0            False               229  CIM,PG,TestCMPI,TestCMPIInstance
test/TestINdSrcNS2             2.40.0            False               229  CIM,PG,TestCMPI,TestCMPIInstance
test/EmbeddedInstance/Static   2.13.0            True                188  PG,CIM
test/TestProvider              2.41.0            False               532  PG,CIM,Test,TEST,TST,TestCMPI,CMPI,cmpiPerf,TestCMPIInstance,LocalizedProvider
test/EmbeddedInstance/Dynamic  2.13.0            True                188  PG,CIM
root/cimv2                     2.41.0            True               1463  CIM,PG,PRS
root                           0.0.0             False                 0
test/cimv2                     0.0.0             False               776  CIM,PRS
test/static                    2.40.0            False               243  CIM,WET,PG

KSchopmeyer commented:
Examples of server schema against OpenPegasus

```
ppp -n ophttp
Enter 'help' for help, <CTRL-D> or ':q' to exit pywbemcli or <CTRL-r> to search history, 
pywbemcli> server schema
/WARNING: Invalid version_qualifier value 2.10, class CIM_DatabaseControlFile.  Modified to [2, 10, 0]
Schema information;
Namespace                      schemas              classes  schema     experimental
                                                             version
-----------------------------  -----------------  ---------  ---------  --------------
root                                                      0
root/PG_InterOp                CIM, PG                  276  2. 40. 0   False
root/PG_Internal               PG                        10  2. 11. 0   False
root/SampleProvider            CIM                      466  2. 41. 0   True
                               CMPI
                               CMPISample
                               CQL
                               CWS
                               PG
                               QExpr
                               Sample
                               TestCMPI
                               no schema
root/benchmark                 CIM, no schema           206  2. 40. 0   False
root/cimv2                     CIM, PG, PRS            1463  2. 41. 0   True
test/CimsubTestNS0             CIM, PG                  212  2. 40. 0   False
test/CimsubTestNS1             CIM, PG                  212  2. 40. 0   False
test/CimsubTestNS2             CIM, PG                  212  2. 40. 0   False
test/CimsubTestNS3             CIM, PG                  212  2. 40. 0   False
test/EmbeddedInstance/Dynamic  CIM, PG                  188  2. 13. 0   True
test/EmbeddedInstance/Static   CIM, PG                  188  2. 13. 0   True
test/TestINdSrcNS2             CIM                      229  2. 40. 0   False
                               PG
                               TestCMPI
                               TestCMPIInstance
                               no schema
test/TestIndSrcNS1             CIM                      229  2. 40. 0   False
                               PG
                               TestCMPI
                               TestCMPIInstance
                               no schema
test/TestProvider              CIM                      532  2. 41. 0   False
                               CMPI
                               LocalizedProvider
                               PG
                               TST
                               Test
                               TestCMPI
                               TestCMPIInstance
                               cmpiPerf
                               no schema
test/WsmTest                   CIM, no schema           182  2. 40. 0   False
test/cimv2                     CIM, PRS                 776  0. 0. 0    False
test/static                    CIM, PG, WET             243  2. 40. 0   False
pywbemcli> server schema --detail
-WARNING: Invalid version_qualifier value 2.10, class CIM_DatabaseControlFile.  Modified to [2, 10, 0]
Schema information, detail;
Namespace                      schemas              classes  schema     experimental
                                                             version
-----------------------------  -----------------  ---------  ---------  --------------
root/PG_InterOp                                         276  2. 40. 0   False
                               CIM                      241  2, 40, 0   False
                               PG                        35  2, 12, 0   False
root/PG_Internal                                         10  2. 11. 0   False
                               PG                        10  2, 11, 0   False
root/SampleProvider                                     466  2. 41. 0   True
                               CIM                      424  2, 41, 0   True
                               CMPI                       1  0, 0, 0    False
                               CMPISample                 3  1, 0, 0    False
                               CQL                        8  2, 2, 0    False
                               CWS                        6  0, 0, 0    False
                               PG                         7  2, 12, 0   False
                               QExpr                      2  2, 2, 0    False
                               Sample                    10  2, 2, 0    False
                               TestCMPI                   1  0, 0, 0    False
                               no schema                  4  2, 2, 0    False
root/benchmark                                          206  2. 40. 0   False
                               CIM                      177  2, 40, 0   False
                               no schema                 29  1, 0, 0    False
root/cimv2                                             1463  2. 41. 0   True
                               CIM                     1382  2, 41, 0   True
                               PG                        20  2, 12, 0   False
                               PRS                       61  2, 8, 0    False
test/CimsubTestNS0                                      212  2. 40. 0   False
                               CIM                      207  2, 40, 0   False
                               PG                         5  2, 12, 0   False
test/CimsubTestNS1                                      212  2. 40. 0   False
                               CIM                      207  2, 40, 0   False
                               PG                         5  2, 12, 0   False
test/CimsubTestNS2                                      212  2. 40. 0   False
                               CIM                      207  2, 40, 0   False
                               PG                         5  2, 12, 0   False
test/CimsubTestNS3                                      212  2. 40. 0   False
                               CIM                      207  2, 40, 0   False
                               PG                         5  2, 12, 0   False
test/EmbeddedInstance/Dynamic                           188  2. 13. 0   True
                               CIM                      183  2, 13, 0   True
                               PG                         5  0, 0, 0    False
test/EmbeddedInstance/Static                            188  2. 13. 0   True
                               CIM                      183  2, 13, 0   True
                               PG                         5  0, 0, 0    False
test/TestINdSrcNS2                                      229  2. 40. 0   False
                               CIM                      207  2, 40, 0   False
                               PG                         1  0, 0, 0    False
                               TestCMPI                  19  0, 0, 0    False
                               TestCMPIInstance           1  0, 0, 0    False
                               no schema                  1  2, 5, 0    False
test/TestIndSrcNS1                                      229  2. 40. 0   False
                               CIM                      207  2, 40, 0   False
                               PG                         1  0, 0, 0    False
                               TestCMPI                  19  0, 0, 0    False
                               TestCMPIInstance           1  0, 0, 0    False
                               no schema                  1  2, 5, 0    False
test/TestProvider                                       532  2. 41. 0   False
                               CIM                      430  2, 41, 0   False
                               CMPI                       3  1, 0, 0    False
                               LocalizedProvider          1  2, 7, 0    False
                               PG                         8  2, 12, 0   False
                               TST                       27  1, 0, 0    False
                               Test                      33  2, 7, 0    False
                               TestCMPI                  20  1, 0, 0    False
                               TestCMPIInstance           1  0, 0, 0    False
                               cmpiPerf                   3  0, 0, 0    False
                               no schema                  6  2, 11, 0   False
test/WsmTest                                            182  2. 40. 0   False
                               CIM                      177  2, 40, 0   False
                               no schema                  5  1, 0, 0    False
test/cimv2                                              776  0. 0. 0    False
                               CIM                      715  0, 0, 0    False
                               PRS                       61  0, 0, 0    False
test/static                                             243  2. 40. 0   False
                               CIM                      177  2, 40, 0   False
                               PG                         2  2, 2, 0    False
                               WET                       64  0, 0, 0    False
pywbemcli> server schema --namespace root/cimv2 --detail
-WARNING: Invalid version_qualifier value 2.10, class CIM_DatabaseControlFile.  Modified to [2, 10, 0]
Schema information, detail;
Namespace    schemas      classes  schema     experimental
                                   version
-----------  ---------  ---------  ---------  --------------
root/cimv2                   1463  2. 41. 0   True
             CIM             1382  2, 41, 0   True
             PG                20  2, 12, 0   False
             PRS               61  2, 8, 0    False
pywbemcli>

KSchopmeyer commented:
1. Schema 0,0,0 should be empty.
2. No space after , on version
3. Schemas should fold after making string rather than as a list
4. (all) or something for the totals
5. no schema to (none)
6. the schema version strinifier is not consistent between schemas and namespacerows.
7. Experimental row True/alse to Experimental/ ""

andy-maier commented on docs/pywbemcli/commands.rst:
```diff
@@ -1556,6 +1560,57 @@ Example:
+The ``server schema`` command returns information on the :ref:`DMTF Schema`
+defined for classes in the namespace()s) of the WBEM server of the
+:term:`current connection`.  It provides information on the maximum value class
+version qualifier (which may represent the DMTF ) in the namespace, whether the
```
"maximum value class version qualifier (which may represent the DMTF)" ???
Also, we should make it clear that each namespace can have multiple schemas.

How about:
"For each schema in each namespace, it provides information on the schema version, whether the schema has experimental classes, and the number of classes in the schema."

KSchopmeyer replied:
DONE. 

However, we are still using schema in two different contexts, CIM Schema (prefix chars on class) and what DMTF releases.

andy-maier replied:
On "CIM schema": At one point, we tried to establish in the DMTF documents that "CIM schema" is the general construct and "CIM Schema" the one from DMTF. This even made it into glossaries.

However, I think this is too subtle and does not work in practice.

Therefore, we could for example agree that in pywbem projects, we use "CIM schema" for the general construct and "DMTF CIM Schema" for the one from DMTF.

KSchopmeyer replied:
Not a problem

andy-maier commented on docs/pywbemcli/commands.rst:
```diff
@@ -1556,6 +1560,57 @@ Example:
+installed in each namespace because the version of the DMTF schema installed is
+not recorded in the installed model so the best that can be done is to find
+the highest value of the Version qualifier that represents the DMTF Schema
+version where the last change was made to that each class.
```
This sounds quite negative. I would phrase it more neutrally:

"Pywbemcli determines the version of a schema by the highest value of the Version qualifier on any of its classes."

KSchopmeyer replied:
DONE.

andy-maier commented on docs/pywbemcli/commands.rst:
```diff
@@ -1556,6 +1560,57 @@ Example:
+Namespace    schemas      classes  schema     experimental
+                                   version
+-----------  ---------  ---------  ---------  --------------
+root/cimv2                     20  2.51.0     Experimental
```
The schemas column i the row with the totals is not consistent with the code, which says "(namespace-info)" there instead of blanks. However, see my comment on that, further down.

KSchopmeyer replied:
Changed per comment below.  Just leaving it, blank is also confusing.

andy-maier commented on docs/appendix.rst:
```diff
@@ -200,6 +200,13 @@ This documentation uses a few special terms to refer to Python types:
       define the use of a query language, it is specified with the string
       ``DMTF:QCL``. CQL is described in DMTF standard :term:`DSP0202`.
 
+   DMTF Schema
```
We need a general term for a schema even if it comes from a vendor and not from the DMTF. The term for that is "CIM schema". Please change to using "CIM schema".

There is a number of uses of the term "DMTF schema" that would need to be adjusted as well.

Finally, the description of this term also needs to be generalized, e.g. in "A schema in the DMTF CIM model"

KSchopmeyer replied:
These are different entities.  the DMTF refers to their releases as CIM schemas also.  We agree that a CIM_schema is, defined by the prefix characters on any given class.  Changed all uses of the use of schema that reference the prefix to CIM Schema.  The appendix is pretty much taken from the specification itself.

andy-maier commented on pywbemtools/pywbemcli/_cmd_server.py:
```diff
@@ -227,6 +229,29 @@ def server_remove_mof(context, **options):
+    """
+    Get information about the server schemas.
+
+    Gets information about the DMTF schemas that define the classes in
```
DMTF schemas -> CIM schemas.

andy-maier commented on pywbemtools/pywbemcli/_cmd_server.py:
```diff
@@ -227,6 +229,29 @@ def server_remove_mof(context, **options):
+    Gets information about the DMTF schemas that define the classes in
+    each namespace. The information provided includes:
+      * The schema version - the maximum version defined in a class
+      * Experimental vs. released schema
```
released -> final

KSchopmeyer replied:
DONE

KSchopmeyer replied:
DONE

andy-maier commented on pywbemtools/pywbemcli/_cmd_server.py:
```diff
@@ -412,3 +437,138 @@ def cmd_server_remove_mof(context, options):
+
+def cmd_server_schema(context, options):
+    """
+    The schema command provides information on the dmtf model schems and
```
dmtf model schema -> CIM schema

KSchopmeyer replied:
DONE

KSchopmeyer replied:
rewrote

andy-maier commented on pywbemtools/pywbemcli/_cmd_server.py:
```diff
@@ -412,3 +437,138 @@ def cmd_server_remove_mof(context, options):
+                if ns != prev_namespace:
+                    prev_namespace = ns
+                    rows.append([ns,
+                                 "(namespace-info)",
```
Since the other rows show the schema name, it is a bit confusing to now talk about "namespace-info". I suggest to show "(all schemas)" or something along these lines

KSchopmeyer replied:
DONE.  I could not think of anything better and a blank is confusing also

andy-maier replied:
In the current code as of commit 6702249, it still says "namespace-info". The suggested "--------" is similarly vague as showing blanks. It is a line with the totals for all schemas in this one namespace, so I could go with any of these (by decreasing priority):
* "(totals)"
* "(all)"
* "(all schemas)"
* "(namespace)"

KSchopmeyer replied:
Slelected (totals) DONE

andy-maier commented on pywbemtools/pywbemcli/_cmd_server.py:
```diff
@@ -412,3 +437,138 @@ def cmd_server_remove_mof(context, options):
+
+            schema_elements = klass.classname.split('_', 1)
+            schema = schema_elements[0] if len(schema_elements) > 1 \
+                else "{no schema}"
```
In the fixed string that denotes all schemas, you have used parenthesis. Here, curly braces are used. I suggest to stick with one pair of characters only, e.g. parenthesis.

KSchopmeyer replied:
I did that specifically so that the no schema entry would be sorted to the bottom of the list.  The alternate was more messy code and it looks better if that is at the bottom of the list.  using parenthesis puts that string at the top of the list.

FIXED

andy-maier commented on pywbemtools/pywbemcli/_cmd_server.py:
```diff
@@ -412,3 +437,138 @@ def cmd_server_remove_mof(context, options):
+                         classes_count,
+                         ns_version_str,
+                         experimental_display(ns_experimental)
+                         ])
```
It is confusing to show a schema version in the non-detailed output, since you are applying the max value on possibly very different version number ranges. I suggest to omit the version column in the non-detailed report.

KSchopmeyer replied:
That is the one data item that this was st

andy-maier replied:
So if a vendor schema version 5.3, and the DMTF CIM schema version is 2.50, then this column in the non-detail case will show 5.3. Does that make sense? I think it does not. A list if different schemas does not have a single version.

KSchopmeyer replied:
Agree, discussion.  Sadly one thing we know is that generally it is the DMTF schema that is the basic driver of versions.   Without the versions, what you see in the non-detail output is for each namespace a list of the schema, the number of classes, and whether there was experimental anywhere in the namespace. No clue about the dmtf schema.  However, you did point out one major error, I am using the max version number as the max dmtf schema version.  I have to isolate that search for max version to classes that are in one of the DMTF schema (CIM, PRS, ???). If someone put a schema version 99.99.00 on a class in their schema I would record that as max in the non-detail view.  That is just a bug.

andy-maier replied:
So you would silently show only the DMTF schema version, ignoring any other schema's versions, without saying so?? In Python there is the principle pof least surprise. Don't think that would be met here...

andy-maier reviewed (changes requested):
See comments

andy-maier commented on pywbemtools/pywbemcli/_common.py:
```diff
@@ -782,6 +782,97 @@ def sort_cimobjects(cim_objects):
+    try:
+        version_list = [int(x) for x in version_str.split('.')]
+    except ValueError:
+        warning_msg("Invalid version_qualifier value {}, class {}. "
```
"version_qualifier" -> "Version qualifier"

KSchopmeyer replied:
DONE

andy-maier commented on pywbemtools/pywbemcli/_common.py:
```diff
@@ -782,6 +782,97 @@ def sort_cimobjects(cim_objects):
+                    "Returned 0.0.0".format(version_str, cln))
+        return [0, 0, 0]
+    except TypeError:
+        warning_msg("Invalid version_qualifier value: {}, class {}. "
```
"version_qualifier" -> "Version qualifier"

KSchopmeyer replied:
DONE

andy-maier commented on pywbemtools/pywbemcli/_common.py:
```diff
@@ -782,6 +782,97 @@ def sort_cimobjects(cim_objects):
+    if len(version_list_mod) > 3:
+        version_list_mod = version_list_mod[:3]
+    warning_msg("Invalid version_qualifier value {}, class {}. "
+                " Modified to {}".format(version_str, cln, version_list_mod))
```
"version_qualifier" -> "Version qualifier"

KSchopmeyer replied:
DONE

KSchopmeyer commented:
Updated per comments except that this leaves us with issue of the phrase CIM schema beiing both the classname prefix and the qualdecls, class release from DMTF.  NOTE: The example is the latest code.

KSchopmeyer commented:
Example of output with OpenPegasus:

```
ppp -n ophttp
Enter 'help' for help, <CTRL-D> or ':q' to exit pywbemcli or <CTRL-r> to search history, 
pywbemcli> server schema
|WARNING: Invalid Version qualifier value 2.10, class CIM_DatabaseControlFile.  Modified to [2, 10, 0]
Schema information; namespaces: all;
Namespace                      CIM                                             classes  schema     experimental
                               schemas                                           count  version
-----------------------------  --------------------------------------------  ---------  ---------  --------------
root                                                                                 0
root/PG_InterOp                CIM, PG                                             276  2.40.0
root/PG_Internal               PG                                                   10
root/SampleProvider            CIM, CMPI, CMPISample, CQL, CWS, PG, QExpr,         466  2.41.0     Experimental
                               Sample, TestCMPI, (no-schema)
root/benchmark                 CIM, (no-schema)                                    206  2.40.0
root/cimv2                     CIM, PG, PRS                                       1463  2.41.0     Experimental
test/CimsubTestNS0             CIM, PG                                             212  2.40.0
test/CimsubTestNS1             CIM, PG                                             212  2.40.0
test/CimsubTestNS2             CIM, PG                                             212  2.40.0
test/CimsubTestNS3             CIM, PG                                             212  2.40.0
test/EmbeddedInstance/Dynamic  CIM, PG                                             188  2.13.0     Experimental
test/EmbeddedInstance/Static   CIM, PG                                             188  2.13.0     Experimental
test/TestINdSrcNS2             CIM, PG, TestCMPI, TestCMPIInstance,                229  2.40.0
                               (no-schema)
test/TestIndSrcNS1             CIM, PG, TestCMPI, TestCMPIInstance,                229  2.40.0
                               (no-schema)
test/TestProvider              CIM, CMPI, LocalizedProvider, PG, TST, Test,        532  2.41.0
                               TestCMPI, TestCMPIInstance, cmpiPerf,
                               (no-schema)
test/WsmTest                   CIM, (no-schema)                                    182  2.40.0
test/cimv2                     CIM, PRS                                            776
test/static                    CIM, PG, WET                                        243  2.40.0
pywbemcli> server schema --detail
\WARNING: Invalid Version qualifier value 2.10, class CIM_DatabaseControlFile.  Modified to [2, 10, 0]
Schema information; detail; namespaces: all;
Namespace                      CIM                  classes  schema     experimental
                               schemas                count  version
-----------------------------  -----------------  ---------  ---------  --------------
root/PG_InterOp                (totals)                 276  2.40.0
                               CIM                      241  2.40.0
                               PG                        35  2.12.0
root/PG_Internal               (totals)                  10
                               PG                        10  2.11.0
root/SampleProvider            (totals)                 466  2.41.0     Experimental
                               CIM                      424  2.41.0     Experimental
                               CMPI                       1
                               CMPISample                 3  1.0.0
                               CQL                        8  2.2.0
                               CWS                        6
                               PG                         7  2.12.0
                               QExpr                      2  2.2.0
                               Sample                    10  2.2.0
                               TestCMPI                   1
                               (no-schema)                4  2.2.0
root/benchmark                 (totals)                 206  2.40.0
                               CIM                      177  2.40.0
                               (no-schema)               29  1.0.0
root/cimv2                     (totals)                1463  2.41.0     Experimental
                               CIM                     1382  2.41.0     Experimental
                               PG                        20  2.12.0
                               PRS                       61  2.8.0
test/CimsubTestNS0             (totals)                 212  2.40.0
                               CIM                      207  2.40.0
                               PG                         5  2.12.0
test/CimsubTestNS1             (totals)                 212  2.40.0
                               CIM                      207  2.40.0
                               PG                         5  2.12.0
test/CimsubTestNS2             (totals)                 212  2.40.0
                               CIM                      207  2.40.0
                               PG                         5  2.12.0
test/CimsubTestNS3             (totals)                 212  2.40.0
                               CIM                      207  2.40.0
                               PG                         5  2.12.0
test/EmbeddedInstance/Dynamic  (totals)                 188  2.13.0     Experimental
                               CIM                      183  2.13.0     Experimental
                               PG                         5
test/EmbeddedInstance/Static   (totals)                 188  2.13.0     Experimental
                               CIM                      183  2.13.0     Experimental
                               PG                         5
test/TestINdSrcNS2             (totals)                 229  2.40.0
                               CIM                      207  2.40.0
                               PG                         1
                               TestCMPI                  19
                               TestCMPIInstance           1
                               (no-schema)                1  2.5.0
test/TestIndSrcNS1             (totals)                 229  2.40.0
                               CIM                      207  2.40.0
                               PG                         1
                               TestCMPI                  19
                               TestCMPIInstance           1
                               (no-schema)                1  2.5.0
test/TestProvider              (totals)                 532  2.41.0
                               CIM                      430  2.41.0
                               CMPI                       3  1.0.0
                               LocalizedProvider          1  2.7.0
                               PG                         8  2.12.0
                               TST                       27  1.0.0
                               Test                      33  2.7.0
                               TestCMPI                  20  1.0.0
                               TestCMPIInstance           1
                               cmpiPerf                   3
                               (no-schema)                6  2.11.0
test/WsmTest                   (totals)                 182  2.40.0
                               CIM                      177  2.40.0
                               (no-schema)                5  1.0.0
test/cimv2                     (totals)                 776
                               CIM                      715
                               PRS                       61
test/static                    (totals)                 243  2.40.0
                               CIM                      177  2.40.0
                               PG                         2  2.2.0
                               WET                       64
pywbemcli> server schema -n root/cimv2
|WARNING: Invalid Version qualifier value 2.10, class CIM_DatabaseControlFile.  Modified to [2, 10, 0]
Schema information; namespaces: root/cimv2;
Namespace    CIM             classes  schema     experimental
             schemas           count  version
-----------  ------------  ---------  ---------  --------------
root/cimv2   CIM, PG, PRS       1463  2.41.0     Experimental
pywbemcli> server schema -n root/cimv2 --detail
/WARNING: Invalid Version qualifier value 2.10, class CIM_DatabaseControlFile.  Modified to [2, 10, 0]
Schema information; detail; namespaces: root/cimv2;
Namespace    CIM          classes  schema     experimental
             schemas        count  version
-----------  ---------  ---------  ---------  --------------
root/cimv2   (totals)        1463  2.41.0     Experimental
             CIM             1382  2.41.0     Experimental
             PG                20  2.12.0
             PRS               61  2.8.0

KSchopmeyer commented:
1.  Change title of schema col to DMTF CIM schema version in non-detailed version
2. Compate row with namespace into first schema row and do not show  the summaries as details
3. Change col to schema version for detailed

KSchopmeyer commented:
Updated per agreement above.

Example output:

```

ppp
Enter 'help' for help, <CTRL-D> or ':q' to exit pywbemcli or <CTRL-r> to search history, 
pywbemcli> server schema
|WARNING: Invalid Version qualifier value 2.10, class CIM_DatabaseControlFile.  Modified to [2, 10, 0]
Schema information; namespaces: all;
Namespace                      Schemas                                         classes  DMTF schema    experimental
                                                                                 count  version
-----------------------------  --------------------------------------------  ---------  -------------  --------------
root                                                                                 0
root/PG_InterOp                CIM, PG                                             276  2.40.0
root/PG_Internal               PG                                                   10
root/SampleProvider            CIM, CMPI, CMPISample, CQL, CWS, PG, QExpr,         466  2.41.0         Experimental
                               Sample, TestCMPI, (no-schema)
root/benchmark                 CIM, (no-schema)                                    206  2.40.0
root/cimv2                     CIM, PG, PRS                                       1463  2.41.0         Experimental
test/CimsubTestNS0             CIM, PG                                             212  2.40.0
test/CimsubTestNS1             CIM, PG                                             212  2.40.0
test/CimsubTestNS2             CIM, PG                                             212  2.40.0
test/CimsubTestNS3             CIM, PG                                             212  2.40.0
test/EmbeddedInstance/Dynamic  CIM, PG                                             188  2.13.0         Experimental
test/EmbeddedInstance/Static   CIM, PG                                             188  2.13.0         Experimental
test/TestINdSrcNS2             CIM, PG, TestCMPI, TestCMPIInstance,                229  2.40.0
                               (no-schema)
test/TestIndSrcNS1             CIM, PG, TestCMPI, TestCMPIInstance,                229  2.40.0
                               (no-schema)
test/TestProvider              CIM, CMPI, LocalizedProvider, PG, TST, Test,        532  2.41.0
                               TestCMPI, TestCMPIInstance, cmpiPerf,
                               (no-schema)
test/WsmTest                   CIM, (no-schema)                                    182  2.40.0
test/cimv2                     CIM, PRS                                            776
test/static                    CIM, PG, WET                                        243  2.40.0
pywbemcli> server schema --detail
|WARNING: Invalid Version qualifier value 2.10, class CIM_DatabaseControlFile.  Modified to [2, 10, 0]
Schema information; detail; namespaces: all;
Namespace                      Schemas              classes  schema     experimental
                                                      count  version
-----------------------------  -----------------  ---------  ---------  --------------
root/PG_InterOp                CIM                      241  2.40.0
                               PG                        35  2.12.0
root/PG_Internal               PG                        10  2.11.0
root/SampleProvider            CIM                      424  2.41.0     Experimental
                               CMPI                       1
                               CMPISample                 3  1.0.0
                               CQL                        8  2.2.0
                               CWS                        6
                               PG                         7  2.12.0
                               QExpr                      2  2.2.0
                               Sample                    10  2.2.0
                               TestCMPI                   1
                               (no-schema)                4  2.2.0
root/benchmark                 CIM                      177  2.40.0
                               (no-schema)               29  1.0.0
root/cimv2                     CIM                     1382  2.41.0     Experimental
                               PG                        20  2.12.0
                               PRS                       61  2.8.0
test/CimsubTestNS0             CIM                      207  2.40.0
                               PG                         5  2.12.0
test/CimsubTestNS1             CIM                      207  2.40.0
                               PG                         5  2.12.0
test/CimsubTestNS2             CIM                      207  2.40.0
                               PG                         5  2.12.0
test/CimsubTestNS3             CIM                      207  2.40.0
                               PG                         5  2.12.0
test/EmbeddedInstance/Dynamic  CIM                      183  2.13.0     Experimental
                               PG                         5
test/EmbeddedInstance/Static   CIM                      183  2.13.0     Experimental
                               PG                         5
test/TestINdSrcNS2             CIM                      207  2.40.0
                               PG                         1
                               TestCMPI                  19
                               TestCMPIInstance           1
                               (no-schema)                1  2.5.0
test/TestIndSrcNS1             CIM                      207  2.40.0
                               PG                         1
                               TestCMPI                  19
                               TestCMPIInstance           1
                               (no-schema)                1  2.5.0
test/TestProvider              CIM                      430  2.41.0
                               CMPI                       3  1.0.0
                               LocalizedProvider          1  2.7.0
                               PG                         8  2.12.0
                               TST                       27  1.0.0
                               Test                      33  2.7.0
                               TestCMPI                  20  1.0.0
                               TestCMPIInstance           1
                               cmpiPerf                   3
                               (no-schema)                6  2.11.0
test/WsmTest                   CIM                      177  2.40.0
                               (no-schema)                5  1.0.0
test/cimv2                     CIM                      715
                               PRS                       61
test/static                    CIM                      177  2.40.0
                               PG                         2  2.2.0
                               WET                       64
pywbemcli> server schema -n root/cimv2
|WARNING: Invalid Version qualifier value 2.10, class CIM_DatabaseControlFile.  Modified to [2, 10, 0]
Schema information; namespaces: root/cimv2;
Namespace    Schemas         classes  DMTF schema    experimental
                               count  version
-----------  ------------  ---------  -------------  --------------
root/cimv2   CIM, PG, PRS       1463  2.41.0         Experimental
pywbemcli> server schema --detail -n root/cimv2
/WARNING: Invalid Version qualifier value 2.10, class CIM_DatabaseControlFile.  Modified to [2, 10, 0]
Schema information; detail; namespaces: root/cimv2;
Namespace    Schemas      classes  schema     experimental
                            count  version
-----------  ---------  ---------  ---------  --------------
root/cimv2   CIM             1382  2.41.0     Experimental
             PG                20  2.12.0
             PRS               61  2.8.0

andy-maier commented on docs/appendix.rst:
```diff
@@ -200,6 +200,24 @@ This documentation uses a few special terms to refer to Python types:
       ``DMTF:QCL``. CQL is described in DMTF standard :term:`DSP0202`.
 
+   Schema
+      A schema (named CIM schema in the DMTF specification :term:`DSP0004`) in
```
In DSP0004 2.8.0, the same terms are used:
* "schema: a set of classes with a single defining authority or owning organization"
* "CIM schema: the schema published by the DMTF that defines the Common Information Model"
Given that, I think we can use these definitions pretty much with the same wording.

KSchopmeyer replied:
Changed to use everywhere schema and CIM schema.

andy-maier commented on docs/appendix.rst:
```diff
@@ -200,6 +200,24 @@ This documentation uses a few special terms to refer to Python types:
+      or owning organization. In the CIM model, the schema elements for a class
+      is defined as the components of the class name separated by the character
+      "_" and the CIM schema is the first such element.  For example ``CIM`` is
+      the schema for the class ``CIM_ManagedElement``.
```
The term "CIM model" is not defined in DSP0004. You use it for the metamodel, but given the DSP0004 definition of "CIM schema" which uses the term "Common Information Model" it can easily be misunderstood to mean the DMTF CIM schema. I suggest not to use the term "CIM model".

KSchopmeyer replied:
DONE

andy-maier commented on docs/changes.rst:
```diff
@@ -42,6 +42,13 @@ Released: not yet
 * Added a 'pywbemlistener' command for running and managing WBEM listeners.
   (issues #430, #479, #948)
 
+* Implement server schema command that returns information about the schema
```
I suggest using plural for "about the schema", since there may be multiple schemas in a namespace.

KSchopmeyer replied:
DONE

andy-maier commented on docs/changes.rst:
```diff
@@ -42,6 +42,13 @@ Released: not yet
+  for each namespace including: 1. The namespace, 2. schema version, 3. whether
+  any classes in the schema/namespace are experimental, 4) the number of
+  classes in this schema, and 5. the DMTF schemas (characters before the `_` in
+  the namespace). (see issue #444)
```
Move item 5 to the front, because the name of the schema is the most important information. Also, I suggest talking about "schema name". The term "DMTF schema" is incorrect because this is about general schemas.

KSchopmeyer replied:
DONE

andy-maier commented on docs/pywbemcli/commands.rst:
```diff
@@ -1559,6 +1563,77 @@ Example:
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``server schema`` command returns information on the  :ref:`Schema`
+defined for classes in the namespace(s) of the :term:`current connection`
```
The phrase "Schema defined for classes" does not describe it well. This focuses on the prefix in the class names, but the schema is more than just the prefix.

I suggest saying "... on the :ref:`schema`s in the namespaces of ..."

KSchopmeyer replied:
DONE

andy-maier commented on docs/pywbemcli/commands.rst:
```diff
@@ -1559,6 +1563,77 @@ Example:
+WBEM server.
+
+The overview reports a summary by namespace of this information and the
+detail view (``server schema --detail`) reports the information for each
```
Missing second back quote after "--detail"

KSchopmeyer replied:
DONE

andy-maier commented on docs/pywbemcli/commands.rst:
```diff
@@ -1559,6 +1563,77 @@ Example:
+detail view (``server schema --detail`) reports the information for each
+:ref:`Schema` in the namespace.
+
+For each CIM schema in each namespace, it provides information on the
```
"CIM schema" -> "schema"

KSchopmeyer replied:
DONE

andy-maier commented on docs/pywbemcli/commands.rst:
```diff
@@ -1559,6 +1563,77 @@ Example:
+value of the Version qualifier on any of the classes in the namespace that
+are in that :ref:`Schema`
+
+For the :ref:`DMTF schema` in the overview report the highest value is
```
"DMTF schema" -> "CIM schema"

KSchopmeyer replied:
DONE

andy-maier commented on docs/pywbemcli/commands.rst:
```diff
@@ -1559,6 +1563,77 @@ Example:
+For the :ref:`DMTF schema` in the overview report the highest value is
+determined by finding the highest version value for any of the DMTF schemas
+(normally 'CIM', or 'PRS'). The highest value anywhere in the namespace should
+be the :ref:`DMTF schema` package that was used to define the qualifier
```
What do you mean with "should be"? Is this a requirement? Or an expectation? Or are we not sure?

KSchopmeyer replied:
Just dropped the sentence

andy-maier commented on docs/pywbemcli/commands.rst:
```diff
@@ -1559,6 +1563,77 @@ Example:
+determined by finding the highest version value for any of the DMTF schemas
+(normally 'CIM', or 'PRS'). The highest value anywhere in the namespace should
+be the :ref:`DMTF schema` package that was used to define the qualifier
+declarations and the CIM/PRS classes in that namespace.
```
Overall, I think the previous paragraphs should be restructured to make some simple statements about the detail mode, and some simple statements about the overview mode, without mixing thee statements.

KSchopmeyer replied:
DONE

andy-maier commented on docs/pywbemcli/commands.rst:
```diff
@@ -1559,6 +1563,77 @@ Example:
+    $ pywbemcli --name myserver server schema
+
+    Schema information; namespaces: all;
+    Namespace                      Schemas              classes  DMTF schema    experimental
```
Given the terms in DSP0004, the column heading "DMTF schema" should be changed to "CIM schema".

KSchopmeyer replied:
DONE

andy-maier commented on docs/pywbemcli/commands.rst:
```diff
@@ -1559,6 +1563,77 @@ Example:
+
+
+The above report would indicate that the namespace root/cimv2 was probably
+created with DMTF released schema version 2.41.0. Other namespaces show a
```
"DMTF released schema version" -> "CIM schema version"

KSchopmeyer replied:
DONE

andy-maier commented on docs/pywbemcli/commands.rst:
```diff
@@ -1559,6 +1563,77 @@ Example:
+
+This report tells a bit more in that it reports that there are classes in the
+root/cimv2 namespace 'CIM' schema that are experimental indicating that the
+DMTF schema used was probably the Version 2.41.0, experimental DMTF schema.
```
I suggest to remove "probably". Even if the classes are a subset of the complete CIM schema, the highest valuf of  a Version qualifier is definitely the CIM schema version, as long as the classes come consistently from a single CIM schema release.

KSchopmeyer replied:
DONE

andy-maier commented on docs/pywbemcli/commands.rst:
```diff
@@ -1614,11 +1689,11 @@ statistics gathering and reporting in pywbemcli:
     The capabilities for managing and retrieving server maintained statistics is
     supported only in some WBEM server implementations. While these capabilities
-    were documented in the DMTF CIM Schema, they were never included as part of
+    were documented in the :ref:`DMTF CIM Schema`, they were never included as part of
```
"DMTF CIM Schema" -> "CIM schema"

KSchopmeyer replied:
DONE

andy-maier commented on pywbemtools/pywbemcli/_cmd_class.py:
```diff
@@ -484,17 +485,23 @@ def class_tree(context, classname, **options):
     """
     Parse a string with 3 positive integers seperated by period (CIM version
-    string) into a 3 integer tuple and return the tuole
+    string) into a 3 integer tuple and return the tuole. Used to parse the
```
Typo: "tuole"

KSchopmeyer replied:
DONE

andy-maier commented on pywbemtools/pywbemcli/_cmd_class.py:
```diff
@@ -484,17 +485,23 @@ def class_tree(context, classname, **options):
     Parse a string with 3 positive integers seperated by period (CIM version
-    string) into a 3 integer tuple and return the tuole
+    string) into a 3 integer tuple and return the tuole. Used to parse the
+    version string as an option
```
Is this sentence incomplete?

KSchopmeyer replied:
DONE

andy-maier reviewed (changes requested):
See comments

KSchopmeyer commented:
Examples after update sunday night

```

 ppp -n ophttp
Enter 'help' for help, <CTRL-D> or ':q' to exit pywbemcli or <CTRL-r> to search history, 
pywbemcli> server schema
|WARNING: Invalid Version qualifier value 2.10, class CIM_DatabaseControlFile.  Modified to [2, 10, 0]
Schema information; namespaces: all;
Namespace                      schemas                                         classes  CIM schema    experimental
                                                                                 count  version
-----------------------------  --------------------------------------------  ---------  ------------  --------------
root                                                                                 0
root/PG_InterOp                CIM, PG                                             276  2.40.0
root/PG_Internal               PG                                                   10
root/SampleProvider            CIM, CMPI, CMPISample, CQL, CWS, PG, QExpr,         466  2.41.0        Experimental
                               Sample, TestCMPI, (no-schema)
root/benchmark                 CIM, (no-schema)                                    206  2.40.0
root/cimv2                     CIM, PG, PRS                                       1463  2.41.0        Experimental
test/CimsubTestNS0             CIM, PG                                             212  2.40.0
test/CimsubTestNS1             CIM, PG                                             212  2.40.0
test/CimsubTestNS2             CIM, PG                                             212  2.40.0
test/CimsubTestNS3             CIM, PG                                             212  2.40.0
test/EmbeddedInstance/Dynamic  CIM, PG                                             188  2.13.0        Experimental
test/EmbeddedInstance/Static   CIM, PG                                             188  2.13.0        Experimental
test/TestINdSrcNS2             CIM, PG, TestCMPI, TestCMPIInstance,                229  2.40.0
                               (no-schema)
test/TestIndSrcNS1             CIM, PG, TestCMPI, TestCMPIInstance,                229  2.40.0
                               (no-schema)
test/TestProvider              CIM, CMPI, LocalizedProvider, PG, TST, Test,        532  2.41.0
                               TestCMPI, TestCMPIInstance, cmpiPerf,
                               (no-schema)
test/WsmTest                   CIM, (no-schema)                                    182  2.40.0
test/cimv2                     CIM, PRS                                            776
test/static                    CIM, PG, WET                                        243  2.40.0
pywbemcli> server schema --detail
WARNING: Invalid Version qualifier value 2.10, class CIM_DatabaseControlFile.  Modified to [2, 10, 0]
Schema information; detail; namespaces: all;
Namespace                      schemas              classes  schema     experimental
                                                      count  version
-----------------------------  -----------------  ---------  ---------  --------------
root/PG_InterOp                CIM                      241  2.40.0
                               PG                        35  2.12.0
root/PG_Internal               PG                        10  2.11.0
root/SampleProvider            CIM                      424  2.41.0     Experimental
                               CMPI                       1
                               CMPISample                 3  1.0.0
                               CQL                        8  2.2.0
                               CWS                        6
                               PG                         7  2.12.0
                               QExpr                      2  2.2.0
                               Sample                    10  2.2.0
                               TestCMPI                   1
                               (no-schema)                4  2.2.0
root/benchmark                 CIM                      177  2.40.0
                               (no-schema)               29  1.0.0
root/cimv2                     CIM                     1382  2.41.0     Experimental
                               PG                        20  2.12.0
                               PRS                       61  2.8.0
test/CimsubTestNS0             CIM                      207  2.40.0
                               PG                         5  2.12.0
test/CimsubTestNS1             CIM                      207  2.40.0
                               PG                         5  2.12.0
test/CimsubTestNS2             CIM                      207  2.40.0
                               PG                         5  2.12.0
test/CimsubTestNS3             CIM                      207  2.40.0
                               PG                         5  2.12.0
test/EmbeddedInstance/Dynamic  CIM                      183  2.13.0     Experimental
                               PG                         5
test/EmbeddedInstance/Static   CIM                      183  2.13.0     Experimental
                               PG                         5
test/TestINdSrcNS2             CIM                      207  2.40.0
                               PG                         1
                               TestCMPI                  19
                               TestCMPIInstance           1
                               (no-schema)                1  2.5.0
test/TestIndSrcNS1             CIM                      207  2.40.0
                               PG                         1
                               TestCMPI                  19
                               TestCMPIInstance           1
                               (no-schema)                1  2.5.0
test/TestProvider              CIM                      430  2.41.0
                               CMPI                       3  1.0.0
                               LocalizedProvider          1  2.7.0
                               PG                         8  2.12.0
                               TST                       27  1.0.0
                               Test                      33  2.7.0
                               TestCMPI                  20  1.0.0
                               TestCMPIInstance           1
                               cmpiPerf                   3
                               (no-schema)                6  2.11.0
test/WsmTest                   CIM                      177  2.40.0
                               (no-schema)                5  1.0.0
test/cimv2                     CIM                      715
                               PRS                       61
test/static                    CIM                      177  2.40.0
                               PG                         2  2.2.0
                               WET                       64
pywbemcli> server schema -n root/cimv2
\WARNING: Invalid Version qualifier value 2.10, class CIM_DatabaseControlFile.  Modified to [2, 10, 0]
Schema information; namespaces: root/cimv2;
Namespace    schemas         classes  CIM schema    experimental
                               count  version
-----------  ------------  ---------  ------------  --------------
root/cimv2   CIM, PG, PRS       1463  2.41.0        Experimental
pywbemcli> server schema --detail -n root/cimv2
|WARNING: Invalid Version qualifier value 2.10, class CIM_DatabaseControlFile.  Modified to [2, 10, 0]
Schema information; detail; namespaces: root/cimv2;
Namespace    schemas      classes  schema     experimental
                            count  version
-----------  ---------  ---------  ---------  --------------
root/cimv2   CIM             1382  2.41.0     Experimental
             PG                20  2.12.0
             PRS               61  2.8.0

KSchopmeyer commented:
Latest round of comments incorporated, repushed and new display in this PR.

andy-maier commented:
Ready to go !
